index.d.ts

```diff
@@ -0,0 +1,36 @@
+import { Model, ModelClass, Page } from 'objection';
+import { QueryBuilder } from 'knex';
+
+declare module 'objection-hashid' {
+  class AuthQueryBuilder<M extends Model, R = M[]> extends Model.QueryBuilder<M, R> {
+    ArrayQueryBuilderType: AuthQueryBuilder<M, M[]>;
+    SingleQueryBuilderType: AuthQueryBuilder<M, M>;
+    NumberQueryBuilderType: AuthQueryBuilder<M, number>;
+    PageQueryBuilderType: AuthQueryBuilder<M, Page<M>>;
+  }
+
+  class HashIdModelClass extends Model {
+    QueryBuilderType: AuthQueryBuilder<this>;
+    QueryBuilder: AuthQueryBuilder<this>;
+
+    static get hashIdSalt(): string;
+
+    static get hashIdMinLength(): number | void;
+
+    static get hashIdAlphabet(): string | void;
+
+    static get hashIdSeps(): string | void;
+
+    static get _hashIdInstance(): string;
+
+    get hashId(): string;
+
+    get hashid(): string;
+
+    static get hashIdField(): string;
+
+    static get hashedFields(): Array<any>;
+  }
+
+  export default function hashid(model: new () => Model): ModelClass<HashIdModelClass>;
```

Comment:
So this is the main issue. For one, it doesn't seem like you can _actually_ use this type from the outside world.

What I had to do is:
1. give the module name `ObjectionHashId`
2. take out the `export default`
3. put `export = ObjectionHashId.hashid;` outside of the module/namespace so that it correctly types what you get when you `require('objection-hashid')`.

However, there _still_ seems to be a type erasure here, as I can't even access properties like `get hashId()` from the returned class (see https://github.com/JaneJeon/objection-hashid/pull/51#pullrequestreview-351290193), let alone being able to call `query()`.

My guess is that `ModelClass<HashIdModelClass>` is not what we want. I've even tried changing the return type of the function as `new () => HashIdModelClass` but to no avail. Can you try experimenting with this and see how you can make the return type have all of the definitions you defined in `HashIdModelClass` and be able to call `.query()` without losing types? Thanks!

Reply:
> What I had to do is:
> 
>     1. give the module name `ObjectionHashId`
> 
>     2. take out the `export default`
> 
>     3. put `export = ObjectionHashId.hashid;` outside of the module/namespace so that it correctly types what you get when you `require('objection-hashid')`.

Can you share this code please? I want to know how you're getting type information from requiring the module vs. importing it.

Reply:
Also, just to be clear, you want the developer to be able to write something like `BaseModel.query().insert().hashId` right?

Reply:
You can get the type information by creating another file in the base directory (say, `test.js`) and importing the module (`const plugin = require('.')`) and creating & using a class from that plugin. The two types to get right are 1. instantiating the class, and 2. calling `.query()` on that class. To check what types you have, you can just turn on `checkJs` with typescript and manually check the test file.

Reply:
Also see https://github.com/JaneJeon/objection-hashid/pull/51#pullrequestreview-351290193

Reply:
Also the `hashId` should just be a property. So for example, if you created an _instance_ of the model, you should be able to access the `hashId`. The `Model.query().someMethod()` is for interacting with the database.

A test code you could use is (don't *actually* run this):
```js
const plugin = require('.')
const { Model } = require('objection-2')

class Post extends plugin(Model) {}
const postInstance = await Post.query().insertAndFetch({})
```

And then you'd need to check that the `Post` model extends off of both Objection's `Model` class and our own plugin model. To do that, you'd need to check the types for Objection-specific methods and properties, such as `Post.tableName`, `Post.jsonSchema`, `Post.query()`, `postInstance.$query()`, `postInstance.$toJSON()`, and the likes, and of course, you'd also need to check types for plugin-specific stuff like `Post.hashIdSeps`, `Post.query().findByHashId()`, `postInstance.hashId`. @geopic

Reply:
> You can get the type information by creating another file in the base directory (say, `test.js`) and importing the module (`const plugin = require('.')`) and creating & using a class from that plugin. The two types to get right are 1. instantiating the class, and 2. calling `.query()` on that class. To check what types you have, you can just turn on `checkJs` with typescript and manually check the test file.

I was asking about how you're getting the information from using `require`, but it's ok, I thought only `import` included type information :man_facepalming: . I've tried other approaches today but sadly no approach has come to fruition so I left a message for the Objection developers [here](https://github.com/Vincit/objection.js/issues/625#issuecomment-582403065) to find out whether it is actually possible to achieve this.

Reply:
If you're using VSCode, you can turn on `checkJs` to true, and then when you hover over any variable (e.g. `Post`) you'll see the actual types, not to mention the "red squiggles" under any mistyped variables.